From 00000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sisyphus <sisyphus@ohmyopencode>
Date: Wed, 5 Aug 2026 00:00:00 +0800
Subject: [PATCH] airoha: add HTTP upgrade support for Nokia XG-040G-MF

Enable CONFIG_CMD_WGET and CONFIG_WGET to allow HTTP-based firmware
upgrade instead of TFTP-only. Add HTTP upgrade commands and bootmenu
entry for upgrading via an HTTP server with DHCP.

Signed-off-by: Sisyphus <sisyphus@ohmyopencode>
---
 configs/an7583_nokia_xg-040g-mf_defconfig | 2 ++
 defenvs/an7583_nokia_xg-040g-mf_env       | 13 +++++++++++++
 2 files changed, 15 insertions(+)

diff --git a/configs/an7583_nokia_xg-040g-mf_defconfig b/configs/an7583_nokia_xg-040g-mf_defconfig
--- a/configs/an7583_nokia_xg-040g-mf_defconfig
+++ b/configs/an7583_nokia_xg-040g-mf_defconfig
@@ -41,6 +41,8 @@ CONFIG_CMD_RARP=y
 CONFIG_CMD_DHCP=y
 CONFIG_CMD_DNS=y
 CONFIG_CMD_PING=y
+CONFIG_CMD_WGET=y
+CONFIG_WGET=y
 CONFIG_CMD_PXE=y
 CONFIG_CMD_CACHE=y
 CONFIG_CMD_UUID=y
diff --git a/defenvs/an7583_nokia_xg-040g-mf_env b/defenvs/an7583_nokia_xg-040g-mf_env
--- a/defenvs/an7583_nokia_xg-040g-mf_env
+++ b/defenvs/an7583_nokia_xg-040g-mf_env
@@ -9,6 +9,9 @@ bootfile=immortalwrt-airoha-an7583-nokia_xg-040g-mf-ubi-initramfs-recovery.itb
 bootfile_bl2=immortalwrt-airoha-an7583-nokia_xg-040g-mf-ubi-preloader.bin
 bootfile_fip=immortalwrt-airoha-an7583-nokia_xg-040g-mf-ubi-bl31-uboot.fip
 bootfile_upg=immortalwrt-airoha-an7583-nokia_xg-040g-mf-ubi-squashfs-sysupgrade.itb
+bootfile_http=immortalwrt-airoha-an7583-nokia_xg-040g-mf-ubi-squashfs-sysupgrade.itb
+httpserverip=192.168.1.254
+httpdstp=80
 bootled_status=green:power
 bootmenu_confirm_return=askenv - Press ENTER to return to menu ; bootmenu 60
 bootmenu_default=0
@@ -22,7 +25,8 @@ bootmenu_3=Load production system via TFTP then write to NAND=setenv noboot 1 ;
 bootmenu_4=[31mLoad BL31+U-Boot FIP via TFTP then write to NAND[0m=run boot_tftp_write_fip ; run bootmenu_confirm_return
 bootmenu_5=[31mLoad BL2 preloader via TFTP then write to NAND[0m=run boot_tftp_write_bl2 ; run bootmenu_confirm_return
 bootmenu_6=Reboot=reset
-bootmenu_7=Reset all settings to factory defaults=run reset_factory ; reset
+bootmenu_7=[34mUpgrade firmware via HTTP[0m=run boot_http_upgrade ; run bootmenu_confirm_return
+bootmenu_8=Reset all settings to factory defaults=run reset_factory ; reset
 boot_first=if button reset ; then led $bootled_status on ; run boot_default ; fi ; bootmenu
 boot_default=run bootcmd ; run boot_tftp_forever
 boot_production=led $bootled_status on ; run ubi_read_production && bootm $loadaddr#$bootconf ; led $bootled_status off
@@ -30,6 +34,10 @@ boot_ubi=run boot_production ; run boot_tftp_forever
 boot_tftp_forever=led $bootled_status on ; while true ; do run boot_tftp ; sleep 1 ; done
 boot_tftp_production=tftpboot $loadaddr $bootfile_upg && env exists replacevol && iminfo $loadaddr && run ubi_write_production ; if env exists noboot ; then else bootm $loadaddr#$bootconf ; fi
 boot_tftp_recovery=tftpboot $loadaddr $bootfile && env exists replacevol && iminfo $loadaddr && run ubi_write_recovery ; if env exists noboot ; then else bootm $loadaddr#$bootconf ; fi
+boot_http_upgrade=wget $loadaddr http://$httpserverip/$bootfile_http && env exists replacevol && iminfo $loadaddr && run ubi_write_production ; if env exists noboot ; then else bootm $loadaddr#$bootconf ; fi
+boot_http_recovery=wget $loadaddr http://$httpserverip/$bootfile && env exists replacevol && iminfo $loadaddr && run ubi_write_recovery ; if env exists noboot ; then else bootm $loadaddr#$bootconf ; fi
+boot_http_write_bl2=mw.b $loadaddr 0xff 0x800 ; setexpr loadaddr_bl2 $loadaddr + 0x800 ; wget $loadaddr_bl2 http://$httpserverip/$bootfile_bl2 && run mtd_write_bl2
+boot_http_write_fip=wget $loadaddr http://$httpserverip/$bootfile_fip && run ubi_write_fip && run reset_factory
 boot_tftp=tftpboot $loadaddr $bootfile && bootm $loadaddr#$bootconf
 boot_tftp_write_bl2=mw.b $loadaddr 0xff 0x800 ; setexpr loadaddr_bl2 $loadaddr + 0x800 ; tftpboot $loadaddr_bl2 $bootfile_bl2 && run mtd_write_bl2
 boot_tftp_write_fip=tftpboot $loadaddr $bootfile_fip && run ubi_write_fip && run reset_factory
